lib/Dialect/FIRRTL/Transforms/LowerLayers.cpp

```diff
@@ -490,6 +490,17 @@ void LowerLayersPass::runOnModuleBody(FModuleOp moduleOp,
         continue;
       }
 
+      // If this operation is a subfield op and it is capturing something
+      // non-passive outside the layerblock, then move this outside the
+      // layerblock.  This avoids creating output ports on the created module.
+      if (isa<SubfieldOp, SubindexOp, SubaccessOp>(op)) {
+        if (llvm::none_of(op.getOperands(),
+                          [&](Value a) { return isAncestor(layerBlock, a); })) {
+          op.moveBefore(layerBlock);
+          continue;
+        }
+      }
```

Comment:
I don't think this does the right thing for subaccess ops when the index is defined inside the layerblock.
```
firrtl.circuit "Foo" {
  firrtl.layer @A bind { }
  firrtl.module @Foo(
    in %a: !firrtl.vector<vector<bundle<a: uint<1>, b flip: uint<2>>, 2>, 2>,
    in %b: !firrtl.uint<1>
  ) {
    firrtl.layerblock @A {
      %n = firrtl.node %b : !firrtl.uint<1>
      %0 = firrtl.subaccess %a[%b] : !firrtl.vector<vector<bundle<a: uint<1>, b flip: uint<2>>, 2>, 2>, !firrtl.uint<1>
    }
  }
}
```
```
./layers.mlir:9:12: error: 'firrtl.matchingconnect' op operand #0 must be a sized passive base type (contains no uninferred widths, or flips) or foreign type, but got '!firrtl.vector<vector<bundle<a: uint<1>, b flip: uint<2>>, 2>, 2>'
      %0 = firrtl.subaccess %a[%b] : !firrtl.vector<vector<bundle<a: uint<1>, b flip: uint<2>>, 2>, 2>, !firrtl.uint<1>
           ^
./layers.mlir:9:12: note: see current operation: "firrtl.matchingconnect"(%0#1, %arg0) : (!firrtl.vector<vector<bundle<a: uint<1>, b flip: uint<2>>, 2>, 2>, !firrtl.vector<vector<bundle<a: uint<1>, b flip: uint<2>>, 2>, 2>) -> ()
// -----// IR Dump After LowerLayers Failed (firrtl-lower-layers) //----- //
"firrtl.circuit"() ({
  "sv.verbatim"() {format_string = "`ifndef layers_Foo_A\0A`define layers_Foo_A", output_file = #hw.output_file<"layers_Foo_A.sv", excludeFromFileList>, symbols = []} : () -> ()
  "firrtl.module"() ({
  ^bb0(%arg2: !firrtl.uint<1>, %arg3: !firrtl.vector<vector<bundle<a: uint<1>, b flip: uint<2>>, 2>, 2>):
    %1 = "firrtl.node"(%arg2) {annotations = [], name = "n", nameKind = #firrtl<name_kind droppable_name>} : (!firrtl.uint<1>) -> !firrtl.uint<1>
    %2 = "firrtl.subaccess"(%arg3, %arg2) : (!firrtl.vector<vector<bundle<a: uint<1>, b flip: uint<2>>, 2>, 2>, !firrtl.uint<1>) -> !firrtl.vector<bundle<a: uint<1>, b flip: uint<2>>, 2>
  }) {annotations = [], convention = #firrtl<convention internal>, layers = [], portAnnotations = [], portDirections = array<i1: false, false>, portLocations = [loc("./layers.mlir":8:12), loc("./layers.mlir":9:12)], portNames = ["b", "a"], portSyms = [], portTypes = [!firrtl.uint<1>, !firrtl.vector<vector<bundle<a: uint<1>, b flip: uint<2>>, 2>, 2>], sym_name = "Foo_A", sym_visibility = "private"} : () -> ()
  "firrtl.module"() ({
  ^bb0(%arg0: !firrtl.vector<vector<bundle<a: uint<1>, b flip: uint<2>>, 2>, 2>, %arg1: !firrtl.uint<1>):
    %0:2 = "firrtl.instance"() {annotations = [], layers = [], lowerToBind, moduleName = @Foo_A, name = "a", nameKind = #firrtl<name_kind droppable_name>, output_file = #hw.output_file<"layers_Foo_A.sv", excludeFromFileList>, portAnnotations = [[], []], portDirections = array<i1: false, false>, portNames = ["b", "a"]} : () -> (!firrtl.uint<1>, !firrtl.vector<vector<bundle<a: uint<1>, b flip: uint<2>>, 2>, 2>)
    "firrtl.matchingconnect"(%0#1, %arg0) : (!firrtl.vector<vector<bundle<a: uint<1>, b flip: uint<2>>, 2>, 2>, !firrtl.vector<vector<bundle<a: uint<1>, b flip: uint<2>>, 2>, 2>) -> ()
    "firrtl.matchingconnect"(%0#0, %arg1) : (!firrtl.uint<1>, !firrtl.uint<1>) -> ()
  }) {annotations = [], convention = #firrtl<convention internal>, layers = [], parameters = [], portAnnotations = [], portDirections = array<i1: false, false>, portLocations = [loc("./layers.mlir":4:8), loc("./layers.mlir":5:8)], portNames = ["a", "b"], portSyms = [], portTypes = [!firrtl.vector<vector<bundle<a: uint<1>, b flip: uint<2>>, 2>, 2>, !firrtl.uint<1>], sym_name = "Foo"} : () -> ()
  "sv.verbatim"() {format_string = "`endif // layers_Foo_A", output_file = #hw.output_file<"layers_Foo_A.sv", excludeFromFileList>, symbols = []} : () -> ()
}) {annotations = [], name = "Foo"} : () -> ()
```

Reply:
For the example, this does error. I will take a look. However, I think the example is supposed to use `%n` as the index and not `%b`?

``` mlir
    firrtl.layerblock @A {
      %n = firrtl.node %b : !firrtl.uint<1>
      %0 = firrtl.subaccess %a[%n] : !firrtl.vector<vector<bundle<a: uint<1>, b flip: uint<2>>, 2>, 2>, !firrtl.uint<1>
    }
```

Reply:
I pushed a change which detects this situation and errors. I'm going to bucket this as improvements to the pass. Actually fixing this is going to be complicated, however.

Does that seem acceptable?